### PR TITLE
[Feat] Activation offloading for distributed lora recipe

### PR DESCRIPTION
**Pull request by Jackmin801**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Change docstring to reflect arg change `activation_checkpoint` -> `enable_activation_checkpoint`
* Add arg `enable_activation_offloading` and logic to do activation offloading

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [x] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by Jackmin801** [comment]
While we were working on this PR, we observed a weird behaviour in the trace.

### Steps to reproduce
#### 1. Clone repo at this branch
```bash
git clone --branch feat-activation-offloading-streams-dist git@github.com:Jackmin801/torchtune
cd torchtune
```
#### 2. Dev Install with torch nightly
```bash
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu121
pip install -e ".[dev]"
```
#### 3. Login to huggingface and download model
```bash
huggingface-cli login --token <HF_TOKEN>
```
```bash
tune download meta-llama/Meta-Llama-3.1-8B-Instruct --output-dir /tmp/Meta-Llama-3.1-8B-Instruct
```
#### 4. Run single device
```bash
#tune run --nnodes 1 --nproc-per-node 2 lora_finetune_distributed --config recipes/configs/llama3_1/8B_lora.yaml \
tune run lora_finetune_single_device --config recipes/configs/llama3_1/8B_lora.yaml \
  optimizer_in_bwd=False \
  enable_activation_offloading=False \
  enable_activation_checkpointing=True \
  compile=False \
  model.lora_attn_modules="[q_proj,v_proj]" \
  model.apply_lora_to_mlp=False \
  model.apply_lora_to_output=False \
  model.lora_rank=8 \
  model.lora_alpha=16 \
  dataset.source=Yukang/LongAlpaca-12k \
  dataset.packed=False \
  dataset.split=train[:10%] \
  dataset.train_on_input=True \
  tokenizer.max_seq_len=8192 \
  metric_logger=torchtune.training.metric_logging.StdoutLogger \
  metric_logger.project=recipe_profiling \
  log_every_n_steps=1 \
  log_peak_memory_stats=True \
  gradient_accumulation_steps=1 \
  max_steps_per_epoch=4 \
  epochs=1 \
  batch_size=2 \
  metric_logger.name=llama3__qlora__seqlen_8192__act_ckpt_True__act_off_True__bs2 \
  profiler.enabled=True \
  profiler.profile_memory=True \
  profiler.with_stack=True \
  profiler.wait_steps=0 \
  profiler.warmup_steps=2 \
  profiler.active_steps=2 \
  profiler.num_cycles=1
```

**Comment by Jackmin801** [comment]
ah. will lint haha. wanted to start hacking first

**Comment by felipemello1** [comment]
Thanks the PR!! Would you be able to run it with activation on and off for ~25-50 steps and compare loss, memory, tps?

If you use weights and biases, its probably the easiest way to take a screenshot (you can login with your gmail, get the token, pip install wandb, wandb login, insert your token)

then you can run your config like this:
`tune run --nproc_per_node 2 lora_finetune_distributed --config llama3_1/8B_lora max_steps_per_epoch=50  epochs=1 metric_logger=torchtune.training.metric_logging.WandBLogger log_peak_memory_stats=True ...other params...`

Please use dataset.packed=True, compile=True, enable_activation_checkpoint=True, enable_activation_offload=true/false

if loss and TPS is about the same, but memory is much lower, this should be easy to approve.

**Comment by Jackmin801** [comment]
image

Loss seems to be roughly the same (difference probably caused by flash attn scheduling). The memory usage is reduced by ~20% with a tps reduction of ~2%. I guess this is ok to merge?

**Comment by felipemello1** [comment]
will get to this tomorrow. Thanks for posting the results!

**Review by felipemello1 — APPROVED** [review]
lgtm! Thanks for the PR!
